pages/docs/concepts/protocol.md

```diff
@@ -0,0 +1,38 @@
+---
+title: Protocol
+weight: 25
+---
+
+
+## What is a protocol?
+A protocol is a mechanism that handles the exchange of messages. It provides a way of getting messages from one point in an event-driven architecture to another. A protocol could either be wired or API.
+
+`Protocol` examples:
+* WebSockets
+* HTTP
+* Kafka
+* MQTT
+
+## Why do we need protocols?
+Whenever a producer detects a state change (events) and publishes those events as messages, a protocol carries those messages to the channel and then to a consumer. Protocol plays a vital role in message transmission.
+
+```mermaid
```

Comment:
this diagram do not visualize what protocol is

maybe better visualize it with examples 🤔 

I mean share some example from different protocols, to "visualize" what protocol is:

#### mqtt

that in MQTT one of rules is that you can specify on protocol level so called quality of service, and you say it is 0 or 1 or 2
```
0 (At most once delivery), 1 (At least once delivery), or 2 (Exactly once delivery)
```
For for example when you set a rule to `2`  for a publisher that sends a message to the broker, you specify how publisher and broker exchange communication to assure the message is delivered only once to the broker.

### http
In HTTP for example you have different request methods. So on the protocol level, it is specified that when you send a request GET to the server, the request should have no body. This means server should not read it, and client should not even allow to send it.

and the protocol describes all these rules, and all apps/libraries must follow. Otherwise there are issues. Just like with diplomatic protocol, of someone do not follow, there is always a diplomatic glitch.

### websocket

another example of protocol that sets a clear rule, before we do stuff on WebSocket protocol level you actually need to connect to the server that supports it, and this connection should be established using HTTP protocol first. And nobody can decide that "I will establish connection using different protocol". I mean, technically sure, but this is against the protocol, so good luck you need to write all tools on your own, and explain to users why you do not follow the protocol 100%

---

now that I wrote above, maybe these examples could be visualized using diagrams 🤔 google, maybe somebody did it already 🤔

Reply:
also good example is POP3, SMTP which are protocols that you maybe heard of, there are core in email exchange. Email servers also need to follow specific protocol when email exchange happens. A set of rules that say how information about the email should be encapsulated. I bet that event sender, recipient are definitions that come from the protocol. Side note: I think I've heard there are people using it for event communication, or different protocol variation.

<img width="661" alt="Screenshot 2022-11-09 at 19 21 06" src="https://user-images.githubusercontent.com/6995927/200910201-9b8dc419-4487-4247-ad0a-4016291758f9.png">

explore, google what to do. Maybe limit to 3 diagrams for 3 selected protocols

I quickly found this for QoS for MQTT. This can be easily represented with mermaid using sequence diagram type afaik.
<img width="1041" alt="Screenshot 2022-11-09 at 19 23 16" src="https://user-images.githubusercontent.com/6995927/200910669-f46542e9-9135-4524-b750-4a570898c4b7.png">